### ISSUE TITLE
Allow Guild Leaders to choose if non-members can post to chat on public guilds

### ISSUE THREAD

**Issue by crookedneighbor** [HabitRPG/habitica#5551]
### When Creating or editing a guild
- [ ] Checkbox on create page, defaulting to not-checked, that says "Prevent Non-Members from sending chat messages - Note: Moderators can post in chat regardless of membership status"
- [ ] Checkbox on edit page that says "Prevent Non-Members from sending chat messages - Note: Moderators can post in chat regardless of membership status"
### On guild view page
- [ ] Members see chat normally (as it is now)
- [ ] if "Prevent Non-Members...chat" flag is not set: Non-members see chat normally
- [ ] if "Prevent Non-Members...chat" flag is set: Non-Members can see chats, but a message about joining the guild is in place of the chat box
### API Level
- [ ] Add check for "Prevent Non-Members...chat" flag in api route for posting a chat message; send appropriate error message
- [ ] Changes to the route must include adjustments/new tests (Ask for help if you need it)
### Misc
- [ ] All new text must use translatable strings

@hairlessbear @GitHubSphinx Thought you might want to see what the admins decided to do.
##

**Comment by crookedneighbor** [comment]
To be clear, anyone is free to work on this ticket, but if no one claims it, I will do it myself. So just make your intention clear if you'd like to tackle this.

As always, I'm available to help.

**Comment by NotGonnaGitUs** [comment]
I'd like to start this but I think I'd need help at the API Level (maybe even before).

**Comment by crookedneighbor** [comment]
:+1: Go for it. Submit a PR that says it's a work in progress when you get stuck and need help.

**Comment by hairlessbear** [comment]
> Add check for "Prevent Non-Members...chat" flag in api route for posting a chat message; send appropriate error message

@crookedneighbor Just to make sure I understand correctly, there's no way this error message could be triggered through normal usage, right? In your description, the chat box isn't there when the flag is set, which (I'm presuming) means that the send chat button isn't there either (or is at least disabled). So the only way this error could be triggered is if a user uses an API call to attempt to post a message in a guild that they aren't a member and that has the flag set, right?

**Comment by crookedneighbor** [comment]
Correct, you'd only encounter the error using the API.

**Comment by hairlessbear** [comment]
@crookedneighbor: In #5509, decoupling notifications and guild membership was discussed. Is that still being considered, or was it decided that it wouldn't happen?

**Comment by crookedneighbor** [comment]
A user would also encounter the error if they loaded guild page in a state where they could post: ie, non-guild members can post, and then the guild leader changes the preference so they cannot. The nonmember would still have the loaded page with the Ability to chat, but the API would keep the chat from posting. 

As far as decoupling notifications, it's not off the table, but not necessary for this ticket.

**Comment by Alys** [comment]
I'm putting this on hold because we're most of the way through a project to change the website's front-end code and there'll be new front-end code that will need to be modified after that.

**Comment by librarianmage** [comment]
Should this still be on hold?

**Comment by Alys** [comment]
To be honest, I'm not too keen on this idea. The public guilds are open to all. Allowing only guild members to post is easily bypassed by joining, posting, then leaving, so it doesn't guarantee that all visible posts will be from current members, it just adds extra inconvenience to a legitimate use case. If the guild leader wants strict control over posting to the guild, they could make a private guild, optionally with a public lobby guild to attract new members.

Even if this is implemented, I don't think it should be done until after we've added an option for not getting notifications from guilds that you've joined. Some people will be interested in a guild, and will be a regular member of the community in that guild, but won't have clicked the join button specifically so that they're not  distracted by notifications of new posts while focussing on their tasks.

**Comment by Alys** [comment]
I'm editing my previous postt to remove my comment about bug reports because "a message about joining the guild is in place of the chat box" would make it very clear what's happening. :) That's a good idea if this is implemented.

**Comment by shanaqui** [comment]
I'm not a fan of this for the reasons Alys mentions -- it seems weird to me that in this case people could still view the guild if they weren't members, as now, but not post unless they join. And if we changed that, people who wanted to enable that might as well just make private guilds.

**Comment by lemoness** [comment]
Agreed with both points. I'm closing this ticket.
